diff --git a/radio.c b/radio.c
index e03d1ed..96e4575 100644
--- a/radio.c
+++ b/radio.c
@@ -35,6 +35,7 @@
 #include "radio.h"
 #include "util.h"
 
+
 static struct {
     char *ident;
     radio_device_t *device;
@@ -291,9 +292,9 @@ void radio_validate_config(const char *filename) {
 
   // First, find out which radio it is.
   while (fgets(line, sizeof(line), conf)) {
-    if (strnstr(line, "Radio:", strlen(line))) {
+    if (strstr(line, "Radio:")) {
       for (i = 0; radio_tab[i].ident; i++) {
-        if (strnstr(line, radio_tab[i].ident, strlen(line))) goto found;
+        if (strstr(line, radio_tab[i].ident)) goto found;
       }
     }
   }
