diff --git a/Makefile b/Makefile
index ade25b4..4e536b8 100644
--- a/Makefile
+++ b/Makefile
@@ -31,6 +31,7 @@ ifeq ($(UNAME),Linux)
 
     # Link libusb statically, when possible
     LIBUSB      = /usr/lib/x86_64-linux-gnu/libusb-1.0.a
+    LDFLAGS     += -lbsd
     ifeq ($(wildcard $(LIBUSB)),$(LIBUSB))
         LIBS    = $(LIBUSB) -lpthread -ludev
     endif
@@ -56,7 +57,7 @@ clean:
 		rm -f *~ *.o core dmrconfig dmrconfig.exe
 
 install:	dmrconfig
-		install -c -s dmrconfig /usr/local/bin/dmrconfig
+		install -c -s dmrconfig $(DESTDIR)/usr/local/bin/dmrconfig
 
 ###
 anytone_ht.o: anytone_ht.c radio.h util.h anytone_ht-map.h
diff --git a/radio.c b/radio.c
index e03d1ed..034919f 100644
--- a/radio.c
+++ b/radio.c
@@ -25,6 +25,7 @@
  * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
  * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
+#include <bsd/string.h>
 #include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
