diff --git a/src/launch/command.cpp b/src/launch/command.cpp
index 94fe239..18b8b2b 100644
--- a/src/launch/command.cpp
+++ b/src/launch/command.cpp
@@ -58,6 +58,10 @@ QList<QString> configBaseDirs() {
 		dir.append("/quickshell");
 	}
 
+	// Extra admin-installed locations, searched *after* user config
+	configDirs.append("/usr/share/quickshell");
+	configDirs.append("/usr/local/share/quickshell");
+
 	return configDirs;
 }
 
