--- navidrome.service.orig	2024-10-26 21:42:56.126801230 +0200
+++ navidrome.service	2024-12-22 13:16:00.750852623 +0100
@@ -11,7 +11,7 @@
 User=navidrome
 Group=navidrome
 Type=simple
-ExecStart=/usr/bin/navidrome --configfile "/etc/navidrome/navidrome.toml"
+ExecStart=/usr/local/bin/navidrome -c /var/lib/navidrome/navidrome.toml -l /var/lib/navidrome/log
 StateDirectory=navidrome
 WorkingDirectory=/var/lib/navidrome
 TimeoutStopSec=20
@@ -47,7 +47,7 @@
 # You can change the following line to `strict` instead of `full` if you don't
 # want navidrome to be able to write anything on your filesystem outside of
 # /var/lib/navidrome.
-ProtectSystem=full
+ProtectSystem=strict
 
 # You can comment the following line if you don't have any media in /home/*.
 # This will prevent navidrome from ever reading/writing anything there.
